File: android_quickbrick_app/src/main/java/com/applicaster/ui/utils/HookExecutor.java
package com.applicaster.ui.utils;

import android.app.Activity;

import androidx.annotation.NonNull;

import com.applicaster.plugin_manager.hook.ApplicationLoaderHookUpI;
import com.applicaster.plugin_manager.hook.HookListener;

import java.util.ArrayDeque;
import java.util.List;
import java.util.Queue;

public class HookExecutor implements HookListener {

    private final Queue<ApplicationLoaderHookUpI> hooks;
    private final Runnable onComplete;
    private final Activity activity;
    private final boolean isAppReady;

    public HookExecutor(@NonNull Activity activity,
                        @NonNull List<ApplicationLoaderHookUpI> hooks,
                        @NonNull Runnable onComplete,
                        boolean isAppReady) {
        this.activity = activity;
        this.hooks = new ArrayDeque<>(hooks);
        this.onComplete = onComplete;
        this.isAppReady = isAppReady;
        next();
    }

    private void next() {
        ApplicationLoaderHookUpI hook = hooks.poll();
        if(null == hook) {
            onComplete.run();
        }
        else {
            if (isAppReady) {
                this.activity.runOnUiThread(() -> hook.executeOnApplicationReady(activity, this));
            } else {
                this.activity.runOnUiThread(() -> hook.executeOnStartup(activity, this));
            }
        }
    }

    @Override
    public void onHookFinished() {
        next();
    }
}

File: android_quickbrick_app/src/main/java/com/applicaster/ui/utils/HookExecutor.kt
package com.applicaster.ui.utils

import android.app.Activity
import com.applicaster.plugin_manager.hook.ApplicationLoaderHookUpI
import com.applicaster.plugin_manager.hook.HookListener
import io.reactivex.CompletableEmitter
import java.util.*

class HookExecutor(private val activity: Activity,
                   hooks: List<ApplicationLoaderHookUpI>,
                   private val onComplete: CompletableEmitter,
                   private val isAppReady: Boolean) : HookListener {

    private val hooks: Queue<ApplicationLoaderHookUpI>

    private fun next() {
        val hook = hooks.poll()
        if (null == hook) {
            onComplete.onComplete()
        } else {
            if (isAppReady) {
                activity.runOnUiThread { hook.executeOnApplicationReady(activity, this) }
            } else {
                activity.runOnUiThread { hook.executeOnStartup(activity, this) }
            }
        }
    }

    override fun onHookFinished() = next()

    init {
        this.hooks = ArrayDeque(hooks)
        next()
    }

}